packages/completer/src/handler.ts

```diff
@@ -474,7 +536,82 @@ export namespace CompletionHandler {
      * it is acceptable for the other methods to be simple functions that return
      * rejected promises.
      */
-    connector: IDataConnector<IReply, void, IRequest>;
+    connector?: IDataConnector<IReply, void, IRequest>;
+    /**
+     * Fetcher for ICompletionItems.
+     * If this is set, it'll be used in lieu of the data connector.
+     */
+    fetchItems?: (
+      request: CompletionHandler.IRequest
+    ) => Promise<CompletionHandler.ICompletionItemsReply>;
+  }
+
+  /**
+   * Type alias for ICompletionItem list.
+   * Implementers of this interface should be responsible for
+   * deduping and sorting the items in the list.
+   */
+  export type ICompletionItems = ReadonlyArray<ICompletionItem>;
+
+  /**
+   * Completion item object based off of LSP CompletionItem.
+   * Compared to the old kernel completions interface, this enhances the completions UI to support:
+   * - differentiation between inserted text and user facing text
+   * - documentation for each completion item to be displayed adjacently
+   * - deprecation styling
+   * - custom icons
+   * and other potential new features.
+   */
+  export interface ICompletionItem {
+    /**
+     * User facing completion.
+     * If insertText is not set, this will be inserted.
+     */
+    label: string;
+
+    /**
+     * Completion to be inserted.
+     */
+    insertText?: string;
+
+    /**
+     * Type of this completion item.
+     */
+    type?: string;
+
+    /**
+     * Image url for icon to be rendered with completion type.
+     */
+    icon?: string;
```

Comment:
This should follow the same pattern as the icon fields in other jlab core interfaces. This is copied out of [`toolbar.tsx`](https://github.com/jupyterlab/jupyterlab/blob/master/packages/apputils/src/toolbar.tsx):

```
    icon?: LabIcon
    iconClass?: string;
    iconLabel?: string;
```